File: 2025/01 January/03 Count Subarrays with given XOR/main.cpp
// Link: https://www.geeksforgeeks.org/problems/count-subarray-with-given-xor/1

#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution
{
public:
    long subarrayXor(vector<int> &arr, int k)
    {
        int n = arr.size(), cnt = 0, XOR = 0;
        unordered_map<int, int> mp;
        mp[0] = 1;
        for (int i = 0; i < n; i++)
        {
            XOR ^= arr[i];
            int temp = XOR ^ k;
            cnt += mp[temp];
            mp[XOR]++;
        }
        return cnt;
    }
};

int main()
{
    Solution object;
    vector<int> arr = {4, 2, 2, 6, 4};
    int k = 6;
    cout << object.subarrayXor(arr, k) << endl;

    arr = {5, 6, 7, 8, 9};
    k = 5;
    cout << object.subarrayXor(arr, k) << endl;

    return 0;
}

File: 2025/01 January/03 Count Subarrays with given XOR/main.py
# Link: https://www.geeksforgeeks.org/problems/count-subarray-with-given-xor/1


class Solution:
    def subarrayXor(self, arr, k):
        n = len(arr)
        cnt = 0
        XOR = 0
        freq_map = {0: 1}

        for i in range(n):
            XOR ^= arr[i]
            temp = XOR ^ k
            if temp in freq_map:
                cnt += freq_map[temp]
            if XOR in freq_map:
                freq_map[XOR] += 1
            else:
                freq_map[XOR] = 1

        return cnt


if __name__ == "__main__":
    object = Solution()
    arr = [4, 2, 2, 6, 4]
    k = 6
    print(object.subarrayXor(arr, k))

    arr = [5, 6, 7, 8, 9]
    k = 5
    print(object.subarrayXor(arr, k))
